websocket/websocket.go

```diff
@@ -0,0 +1,152 @@
+package websocket
+
+import (
+	"errors"
+	"log"
+	"net/url"
+	"os"
+	"os/signal"
+	"time"
+
+	"github.com/gorilla/websocket"
+)
+
+// Client for forex api instance
+type Client struct {
+	Address   *string
+	Conn      *websocket.Conn
+	Listeners *[]func(interface{})
+	logger    *log.Logger
+}
+
+// New : create the weboscket instance
+func New(address *string, logger *log.Logger) (*Client, error) {
+	if address == nil {
+		return nil, errors.New("'address' is missing")
+	}
+	if logger == nil {
+		logger = log.New(os.Stderr, "WebSocket: ", log.LstdFlags)
+	}
+	return &Client{
+		Address:   address,
+		logger:    logger,
+		Listeners: &[]func(interface{}){},
+	}, nil
+}
+
+// Connect to websocket endpoint
+func (client *Client) Connect(query string) error {
+	interrupt := make(chan os.Signal, 1)
+	signal.Notify(interrupt, os.Interrupt)
+
+	// connection
+	u := url.URL{
+		Scheme:   "ws",
+		Host:     *client.Address,
+		Path:     "",
+		RawQuery: query,
+	}
+	client.logger.Printf("connecting to %s", u.String())
+
+	dialer := websocket.DefaultDialer
+	dialer.ReadBufferSize = 1024
+	dialer.WriteBufferSize = 1024
+
+	c, _, err := dialer.Dial(u.String(), nil)
+	if err != nil {
+		return err
+	}
+	client.Conn = c
+
+	// reader
+	done := make(chan bool)
+	go func() {
+		defer close(done)
+		for {
+			var message interface{}
+			err := c.ReadJSON(&message)
+			if err != nil {
+				client.logger.Println("error read:", err)
+				return
+			}
+			go func() {
+				for _, listener := range *client.Listeners {
+					go listener(message)
+				}
+			}()
+		}
+	}()
+
+	// states handler
+	go func() {
+		for {
```

Comment:
should not be a loop

Reply:
removed os interrupt command. the loop will be gone.